model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/PersistentSessionsWorker.java

```diff
@@ -81,41 +86,60 @@ private void process(ArrayBlockingQueue<PersistentUpdate> queue) throws Interrup
             if (polled != null) {
                 batch.add(polled);
                 queue.drainTo(batch, maxBatchSize - 1);
-                try {
-                    LOG.debugf("Processing %d deferred session updates.", batch.size());
-                    Retry.executeWithBackoff(iteration -> {
-                                if (iteration < 2) {
-                                    // attempt to write whole batch in the first two attempts
-                                    KeycloakModelUtils.runJobInTransaction(factory,
-                                            innerSession -> batch.forEach(c -> c.perform(innerSession)));
-                                    batch.forEach(PersistentUpdate::complete);
-                                } else {
-                                    LOG.warnf("Running single changes in iteration %d for %d entries", iteration, batch.size());
-                                    ArrayList<PersistentUpdate> performedChanges = new ArrayList<>();
-                                    List<Throwable> throwables = new ArrayList<>();
-                                    batch.forEach(change -> {
-                                        try {
-                                            KeycloakModelUtils.runJobInTransaction(factory,
-                                                    change::perform);
-                                            change.complete();
-                                            performedChanges.add(change);
-                                        } catch (Throwable ex) {
-                                            throwables.add(ex);
+                KeycloakModelUtils.runJobInTransaction(factory, outerSession -> {
+                    TracingProvider tracing = outerSession.getProvider(TracingProvider.class);
+                    Tracer process = tracing.getTracer("PersistentSessionsWorker");
+                    SpanBuilder spanBuilder = process.spanBuilder("PersistentSessionsWorker.process");
+                    if (batch.size() > 1) {
+                        batch.stream().map(update -> update.getSpan().getSpanContext()).forEach(spanBuilder::addLink);
+                    } else {
+                        spanBuilder.setParent(Context.current().with(batch.get(0).getSpan()));
+                    }
+                    Span span = tracing.startSpan(spanBuilder);
+                    try {
+                        if (batch.size() > 1) {
+                            batch.forEach(persistentUpdate -> {
+                                persistentUpdate.getSpan().addLink(span.getSpanContext());
+                            });
+                        }
```

Comment:
I think it's not necessary - not even preferred to add links to already created spans[1]. There were quite a lot of discussions about using links before/after creation[2].

[1] https://github.com/open-telemetry/opentelemetry-java/blob/e78de66e5fe345c9cc2b1aee5d42978c38925ca4/api/all/src/main/java/io/opentelemetry/api/trace/Span.java#L372
[2] https://github.com/open-telemetry/opentelemetry-specification/issues/454

Reply:
I can check again in the UI how this then works. What I want to achieve: 

1. I want to link from the request to the batch processing (this is this code)
2. I want to link from the batch processing to the request (this is where I update the spanBuilder above)

When I leave out the first bit, my individual requests don't link to the batch, which makes it hard to analyze. 

I see that there has been a discussion, but now it is re-allowed. 

I pushed one more commit which instead of adding links to an existing span, it creates a new span and adds the links to that new span in the builder phase. 

You can test this by locally by changing the `> 1`  to `> 0`. It will then create the following output in Jaeger: 

![image](https://github.com/user-attachments/assets/1f2a17c1-6940-4336-be85-63eaa77bec46)

I didn't find how to change the reference type using the API. Still it should suffice. 

Let me know what you think about this approach. 

cc: @brunobat

Reply:
Links should be used to link to related spans not having a parent/child relationship like async independent tasks spawning from an http request, assuming they will not be needed for the user's response.
Assuming this is the case, links are ok. About the link must be added upon creation of the span, that's related with sampling that was changed, like Alexander mentions. It is now experimental, however, I don't see them removing that again, given the previous community pushback.

Reply:
In this case this is synchronous, but with batching we have a relation of one span having multiple parents. That's then the moment where I use linking. I assume this is still ok to us it here.